libqtile/widget/widgetbox.py

```diff
@@ -0,0 +1,169 @@
+from libqtile.widget import base
+
+
+class WidgetBox(base._Widget):
+    """A widget to declutter your bar.
+
+    WidgetBox is a widget that hides widgets by default but shows them when
+    the box is opened.
+
+    Widgets that are hidden will still update etc. as if they were on the main
+    bar.
+
+    Button clicks are passed to widgets when they are visible so callbacks will
+    work.
+
+    Widgets in the box alsos remain accessible via command interfaces.
+
+    Widgets can only be added to the box via the configuration file. The widget
+    is conigured by adding widgets to the "widgets" parameter as follows:
+
+        widget.WidgetBox(widgets=[
+            widget.TextBox(text="This widget is in the box"),
+            widget.Memory(),
+            ]),
+    """
+    orientations = base.ORIENTATION_HORIZONTAL
+    defaults = [
+        (
+            "font",
+            "sans",
+            "Text font"
+        ),
+        (
+            "fontsize",
+            None,
+            "Font pixel size. Calculated if None."
+        ),
+        (
+            "fontshadow",
+            None,
+            "font shadow color, default is None(no shadow)"
+        ),
+        (
+            "foreground",
+            "#ffffff",
+            "Foreground colour."
+        ),
+        (
+            "close_button_location",
+            "left",
+            "Location of close button when box open ('left' or 'right')"
+        ),
+        (
+            "text_closed",
+            "[<]",
+            "Text when box is closed"
+        ),
+        (
+            "text_open",
+            "[>]",
+            "Text when box is open"
+        ),
+    ]
+
+    def __init__(self, widgets=list(), **config):
+        base._Widget.__init__(self, bar.CALCULATED, **config)
+        self.add_defaults(WidgetBox.defaults)
+        self.box_is_open = False
+        self.widgets = widgets
+        self.add_callbacks({"Button1": self.toggle_box})
+
+        if self.close_button_location not in ["left", "right"]:
+            val = self.close_button_location
+            msg = "Invalid value for 'close_button_location': {}".format(val)
+            logger.warning(msg)
+            self.close_button_location = "left"
+
+    def _configure(self, qtile, bar):
+        base._Widget._configure(self, qtile, bar)
+        self.set_box_label()
+
+        for idx, i in enumerate(self.widgets):
+            if i.configured:
+                i = i.create_mirror()
+                self.widgets[idx] = i
+            qtile.register_widget(i)
+            i._configure(qtile, bar)
+
+    def calculate_length(self):
+        return self.layout.width
+
+    def set_box_label(self):
+        self.layout = self.drawer.textlayout(
+            self.text_open if self.box_is_open else self.text_closed,
+            self.foreground,
+            self.font,
+            self.fontsize,
+            self.fontshadow,
+            markup=False,
+        )
+
```

Comment:
Currently you're creating a new textlayout each time `self.toggle` is called. If the only thing that needs updating is the text, you can update that directly by doing `self.layout.text = self.text_open if self.box_is_open else self.text_closed`